MH-12814 Manually Select And Upload Thumbnails

krinnewitz opened the pull request:
This patch adds a new tab "Thumbnail" to the video editor which provides
the following new functionality:

- Upload a new thumbnail
- Go back to the default thumbnail (extracted at fixed position)
- Extract a thumbnail from the video at the current playback position
- Show a preview of the thumbnail

All involved flavors and tags are configurable. The External API and
OAI-PMH channels are automatically updated when the thumbnail changes.
To update other channels, you have to use the 'publish' workflow.
Dual-stream videos are supported as well. However, the default thumbnail
is only created for one of the tracks.

To update the external API and OAI-PMH on-the-fly, new code was added to
replace media package elements without retracting and republishing the
whole media package.

This first version is a bit slow since it waits for all channels to be
updated. This may not be necessary and could probably be optimized further.

There already was an "Asset Upload" mechanism implemented in Opencast,
which already allowed to upload a new thumbnail. This was left untouched
and should still work.

Here are some screenshots of the new "Thumbnail" tab in the video editor:

Default Thumbnail in use:

![thumbnails_default](https://user-images.githubusercontent.com/1590263/39353957-75f4adaa-4a09-11e8-9461-3b98a930db54.png)

Extracted thumbnail (single track):

![thumbnails_from_single_track](https://user-images.githubusercontent.com/1590263/39353958-761099d4-4a09-11e8-8d30-4ad7910f07bc.png)

Extracted thumbnail (dual track)

![thumbnails_from_dual_track](https://user-images.githubusercontent.com/1590263/39353959-762abf44-4a09-11e8-8271-d21da97645ca.png)

Please note: This PR was rebased to include PR #193 since it is needed to access some thumbnail-related settings when running workflows. Please discuss any code regarding PR #193 by commenting on PR #193  instead of this one.

Also, due to a bug in develop, you won't be able to test publishing to engage, unless you manually merge-in PR #195.

This work is sponsored by SWITCH.

Commit:
MH-12814 Manually Select And Upload Thumbnails

This patch adds a new tab "Thumbnail" to the video editor which provides
the following new functionality:

- Upload a new thumbnail
- Go back to the default thumbnail
- Extract a thumbnail from the video at the current playback position

All involved flavors and tags are configurable. The External API and
OAI-PMH channels are automatically updated when the thumbnail changes.
To update other channels, you have to use the 'publish' workflow.
Dual-stream videos are supported as well. However, the default thumbnail
is only created for one of the tracks.

To update the external API and OAI-PMH on-the-fly, new code was added to
replace media package elements without retracting and republishing the
whole media package.

This first version is a bit slow since it waits for all channels to be
updated. This may not be necessary and could probably be optimized further.

There already was an "Asset Upload" mechanism implemented in Opencast,
which already allowed to upload a new thumbnail. This was left untouched
and should still work.

This patch extends the DistributionServices to also support synchronous
distribution and retraction of elements which speeds things up a lot
since the whole job dispatching logic with its wait intervals can be
bypassed. However, it should be kept in mind that this functionality
should only be used in rare cases since those distribution workloads
cannot be load balanced.

Co-authored-by: Philipp Middendorf <middendorf@plapadoo.de>
Co-authored-by: Sven Stauber <sven.stauber@switch.ch>

staubesv commented:
@krinnewitz Please note that his PR currently has merge conflicts.

staubesv commented:
@krinnewitz Absolutely love the synchronous distribution!!!

staubesv commented:
@krinnewitz Please note that this PR has merge conflicts. Can you please resolve them?

krinnewitz commented:
Thanks for the hint, fixed.

staubesv commented:
Hi @krinnewitz  Seems the documentation changes contain the wrong unordered list items:

```
./admin/docs/configuration/workflow.md: 336: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
./admin/docs/configuration/workflow.md: 337: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
./admin/docs/configuration/workflow.md: 338: MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
./admin/docs/configuration/workflow.md: 336: MD006/ul-start-left Consider starting bulleted lists at the beginning of the line [Expected: 0; Actual: 2]
./admin/docs/configuration/workflow.md: 336: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
./admin/docs/configuration/workflow.md: 355: MD013/line-length Line length [Expected: 120; Actual: 214]

```

Sorry, have forgot about those checks ;-) Can you please fix it?

krinnewitz commented:
Sure, done :)

staubesv commented:
Hi @krinnewitz I've just merged #264 which caused merge conflicts in this PR. Could you please resolve them?

krinnewitz commented:
Hi @staubesv thanks. Just rebased and fixed merge conflicts.

staubesv commented on etc/security/mh_default_org.xml:
```diff
@@ -48,6 +48,7 @@
     <sec:intercept-url pattern="/admin-ng/event/*/asset/publication/*.json" method="GET" access="ROLE_ADMIN, ROLE_UI_EVENTS_DETAILS_ASSETS_VIEW" />
     <sec:intercept-url pattern="/admin-ng/event/catalogAdapters" method="GET" access="ROLE_ADMIN, ROLE_UI_EVENTS_DETAILS_METADATA_VIEW" />
     <sec:intercept-url pattern="/admin-ng/event/events.json" method="GET" access="ROLE_ADMIN, ROLE_UI_EVENTS_VIEW" />
+    <sec:intercept-url pattern="/admin-ng/event/workflowProperties" method="POST" access="ROLE_ADMIN, ROLE_UI_EVENTS_CREATE, ROLE_UI_TASKS_CREATE, ROLE_UI_EVENTS_EDITOR_VIEW" />
```
Also needs

     <sec:intercept-url pattern="/admin-ng/tools/*/thumbnail.json" method="POST" access="ROLE_ADMIN, ROLE_UI_EVENTS_EDITOR_EDIT" />

krinnewitz replied:
Right, thanks, fixed ;)

gregorydlogan commented:
In testing this today, I can see that it's updating the thumbnail, but not correctly.  STR:

- Open the editor
- Jump in somewhere in the video
- Select the new thumbnail from the video.  I was using the right stream, but I'd imagine the left is affected as well
- See the correct thumbnail in the preview (ie, in the editor)
- Select publish as the workflow, then save & continue
- Wait

The thumnail appears to be somewhere else in the video in my testing.  If I had to guess, it's somewhere near the first detected transition, rather than the arbitrary point I chose.  Upon loading the editor again and going back into the thumbnail editor, the *correct* thumbnail is displayed, however the incorrect version is present in the search index.

Finally, is there a faster way to update the thumbnail than reprocessing the entire workflow?  The default workflow does the full reencode, which seems very heavy for just extracting a thumbnail...

krinnewitz commented:
Hi @gregorydlogan. Thanks for testing.

If your event ist already published, you don't have to run the workflow after changing the thumbnail. The thumbnail will be updated automatically in OAI-PMH and external API without any workflows involved. 

However, running the workflow should extract the same thumbnail as displayed in the video editor. I will soon re-test this using your STR.

Update: I justed checked and it seems to work for me. Please note, that the engage-player uses a different thumbnail. This one is not changed when updating the thumbnaiil in the video editor. The thumbnail changed in the video editor is only published to OAI-PMH and external API.

lkiesow commented:
> If your event ist already published, you don't have to run the workflow after changing the thumbnail. The thumbnail will be updated automatically in OAI-PMH and external API without any workflows involved.

Is that the case even if auto-republication is turned off (which should be the default)? That shouldn't happen.

gregorydlogan commented:
@krinnewitz I'm assuming changing the engage player's thumbnail is out of scope?  It seems like this could generate a fair amount of confusion if it's using different images for the same outputs in different places.

krinnewitz commented:
@gregorydlogan I just noticed that with the default configuration, the engage thumbnails should actually be updated as well (falvor is "*/search+preview" and tag "engage-download" is applied). I will have to investigate why this doesn't work.

@lkiesow you mean this?
``` 
# whether to automatically propagate meta data changes of episodes to OAI-PMH
propagate.episode=false
```

staubesv commented:
Hi @krinnewitz Please note that https://github.com/opencast/opencast/pull/314 has caused conflicts in this PR. Could you please fix them?

krinnewitz commented:
@staubesv conflict fixed.

gregorydlogan commented:
@krinnewitz I'm assuming those pushes don't resolve the engage thumbnail issue?

staubesv commented:
@gregorydlogan Please note that this pull requests includes https://github.com/opencast/opencast/pull/193 and cannot be merged before #193 is merged. Our focus is currently on #193. Once that passes, there will be some more updates to this PR.

Considering incremental publication: As we do not use the search index, we expect that whoever would want to use "incremental" updates of thumbnails for the search index directly would provide an implementation for this. In the standard configuration, I would recommend to disable the automatic distribution of thumbnails completely (it is always enabled for the video editor itself, though). While it is true that this makes changing thumbnails for published events expensive, it still works fine for the main use case (cutting video, select thumbnail and then publish).

karendolan commented:
@staubesv  are you saying this pull does not allow you to choose a thumbnail from the trim UI that can be pubished to Engage players? Or just that it has to be tagged differently (i.e. a tag operation has to be added before the publish operation in the workflow)?

staubesv commented:
@karendolan It can be published to the Engage player, too. This requires the start of a workflow, though. For the publication channels of OAI-PMH and the External API this can be avoided by enabling the auto-redistribution.

gregorydlogan commented on docs/guides/admin/docs/configuration/workflow.md:
```diff
@@ -316,6 +316,82 @@ Example:
+### Uploaded Thumbnail
+
+The Admin UI supports the upload of an image to be used as thumbnail. This thumbnail image will available to
+workflows for further processing as media package attachment with a configurable flavour
```
As much as I hate it, we're using the american version `flavor` everywhere.  Don't worry about fixing it here, there are a couple of other places in the docs where this version is used.  I'll do a bulk fix once this has been merged.

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:
```diff
@@ -370,14 +406,131 @@ public Response getVideoEditor(@PathParam("mediapackageid") final String mediaPa
+          .map(MediaPackageElement::getFlavor)
+          .map(e -> {
+            String side = null;
+            if (e.equals(this.adminUIConfiguration.getSourceTrackLeftFlavor())) {
```
What is `e` doesn't equal either of these?

krinnewitz replied:
I am not sure if there are any configurations in which you may have more than 2 tracks. Just to be extra cautious, I have now set `side` to `"unknown"` in case the track doesn't have the left, nor the right source track flavor.

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:
```diff
@@ -370,14 +406,131 @@ public Response getVideoEditor(@PathParam("mediapackageid") final String mediaPa
+              f("defaultPosition", thumbnail.getDefaultPosition()),
+              f("type", ThumbnailImpl.ThumbnailSource.UPLOAD.name()),
+              f("url", distElement.getURI().toString())))));
+        } else if (current.isFormField() && "TRACK".equalsIgnoreCase(current.getFieldName())) {
```
Are `TRACK`, `FILE`, `POSITION`, and `DEFAULT` defined as a constants somewhere?

krinnewitz replied:
Good point, I just introduced constants here.

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:
```diff
@@ -370,14 +406,131 @@ public Response getVideoEditor(@PathParam("mediapackageid") final String mediaPa
+          }
+        } else if (current.isFormField() && "POSITION".equalsIgnoreCase(current.getFieldName())) {
+          final String value = Streams.asString(current.openStream());
+          position = OptionalDouble.of(Double.parseDouble(value));
```
If we're looking for something in the set (0, 1, 2), shouldn't we be using Integer to do the parsing?

krinnewitz replied:
The value sent by toolsResource.js for `position` is actually a double

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java:
```diff
@@ -824,7 +1007,17 @@ static EditingInfo parse(JSONObject obj) {
 
-      return new EditingInfo(segments, tracks, Opt.nul((String) obj.get("workflow")));
+      OptionalDouble defaultThumbnailPosition = OptionalDouble.empty();
+      final Object defaultThumbnailPositionObj = obj.get("defaultThumbnailPosition");
```
This seems like it could be a constant too?

krinnewitz replied:
Yep, done!

gregorydlogan commented on modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/ThumbnailImpl.java:
```diff
@@ -0,0 +1,494 @@
+    return Arrays.stream(mp.getPublications()).filter(p -> p.getChannel().equalsIgnoreCase(channelId)).findAny();
+  }
+
+  private Publication replaceIgnoreExceptions(final MediaPackage mp, final String channelId,
```
Method name should be altered to match what it's doing (ie, it's not ignoring the exceptions, it's passing them upwards).

krinnewitz replied:
Actually, the method is not needed anymore. I removed it.

staubesv committed:
Fix thumbnail preview generation

The workflow that generates the thumbnail preview images for the video editor needs to consider the type of the thumbnail to generate the correct preview image

staubesv committed:
Publish the correct default thumbnail

The default thumbnail needs to be extracted from the uncut media as thumbnailPosition refers to that

staubesv committed:
Fix publication of search previews

The default thumbnail is taken from the presenter track if available and from the presentation track otherwise.

staubesv commented:
Hi @gregorydlogan and @krinnewitz 

Finally found some time to work on this. https://github.com/plapadoo/opencast/pull/25 fixes some problems in the workflows:
- Update Preview now creates the correct thumbnail previews for the video editor
- Publish now correctly publishes the thumbnail to the media module 

I will provide another PR that allows the auto-distribution to OAI-PMH and External API channels to be disabled.

gregorydlogan commented:
I have a pending review with a few minor comments as well.  Should be done before lunch.

gregorydlogan commented on modules/publication-service-oaipmh/src/main/java/org/opencastproject/publication/oaipmh/endpoint/OaiPmhPublicationRestService.java:
```diff
@@ -68,6 +72,8 @@
 
   /** The logger */
   private static final Logger logger = LoggerFactory.getLogger(OaiPmhPublicationRestService.class);
+  private static final String SEPARATOR = ";;";
```
I've seen this constant in a couple of places.  There appears to be a number of OAIPMH related bundles, does it make sense to make a constants class somewhere to hold this (and any other common constants)?

krinnewitz replied:
For now, I have moved the constant to the `OaiPmhPublicationService` which at least avoids the redundancy. In general, I think serializing arrays this way is quite hacky. Maybe we should have a look at all the places where arrays are serialized to strings in opencast and create a common solution. However, I would like to not increase the size of this PR more than necessary. So maybe someone can fix the array serialization separately :)

gregorydlogan reviewed (commented):
Minor nits, but basically ready to go.  Will wait for the other changes as mentioned by @staubesv .

staubesv committed:
Handle events without thumbnail information gracefully

Events before introducing thumbnail support don't have a thumbnail type which caused an exception in the front-end.

staubesv committed:
Sign URLs to thumbnail previews

If Stream Security is enabled, the URLs of the thumbnail preview images need to be signed. Otherwise, the video editor cannot load them.

staubesv committed:
Add configuration property to enable/disable auto distribution

Not everybody necessarily want the thumbnail images to be automatically created and distributed.

staubesv commented:
@gregorydlogan and @krinnewitz Just added https://github.com/plapadoo/opencast/pull/26/files with some bugfixes from our codebase.

Commit:
Fix review findings

krinnewitz commented:
Thanks  @gregorydlogan and @staubesv. I have just merged the PRs by @staubesv, fixed the findings by @gregorydlogan, and rebased everything onto current develop. 

@gregorydlogan I have used separate commits for all the changes since your review, so you don't have to review the whole PR again :D

When merging, all the commits may be squashed.

KatrinIhler commented on docs/guides/admin/docs/configuration/workflow.md:
```diff
@@ -316,6 +316,82 @@ Example:
+
+If you need another flavor for the thumbnail in your specific environment, be sure to adapt the value of
+`thumbnail.source.flavor.type` and `thumbnail.source.flavor.subtype` in `etc/org.opencastproject.adminui.cfg`
+so that the Admin UI addes the thumbnail images with the expected flavour.
```
Typo in _addes_

krinnewitz replied:
Thanks, fixed.

KatrinIhler commented on etc/workflows/partial-publish.xml:
```diff
@@ -10,6 +10,15 @@
+    <operation
+        id="analyze-tracks"
+        exception-handler-workflow="partial-error"
+        description="Analyze tracks in media package an set control variables">
```
Typo in _an_

krinnewitz replied:
Thanks, fixed.

KatrinIhler reviewed (commented):
Just some nitpicks, test went fine!

Commit:
Fix typos

krinnewitz commented:
Thanks @KatrinIhler. Typos fixed.
